src/base58.rs

```diff
@@ -0,0 +1,441 @@
+    fn to_base58(&self) -> String {
+        let mut zero_counter: usize = 0;
+        let mut encoding_flag = false;
+
+        let mut b58_bytes: Vec<u32> = Vec::new();
+
+        let mut first_b58_byte_iter = true;
+        let mut size = 0;
+        for &byte in self.iter() {
+            if !encoding_flag {
+                zero_counter += (byte == 0) as usize;
+                if byte == 0 {
+                    continue;
+                }
+                encoding_flag = true;
+            }
+
+            let mut carry = byte as u32;
+            let mut i = 0;
+
+            while carry != 0 || i < size {
+                if first_b58_byte_iter {
+                    b58_bytes.push(0);
+                    first_b58_byte_iter = false;
+                } else {
+                    if let Some(b58_byte) = b58_bytes.get(i) {
+                        carry += b58_byte << 8;
+                    } else {
+                        b58_bytes.push(0);
+                    }
+                }
+
+                b58_bytes[i] = carry % 58;
+                carry /= 58;
+
+                i += 1;
+            }
+
+            size = i;
+        }
+
+        let encoding_length = zero_counter as usize + b58_bytes.len();
+        let mut b58_encoding: Vec<u8> = vec![b'1'; encoding_length];
+
+        for (i, &b58_byte) in b58_bytes.iter().rev().enumerate() {
+            b58_encoding[zero_counter + i] = ALPHABET[b58_byte as usize];
+        }
+
+        String::from_utf8(b58_encoding).unwrap()
+    }
+}
+
+impl FromBase58 for str {
+    fn from_base58(&self) -> Result<Vec<u8>, FromBase58Error> {
+        let mut raw_bytes: Vec<u8> = Vec::new();
+        let mut final_bytes: Vec<u8> = Vec::new();
+
+        let mut encoding_flag = false;
+
+        let mut first_byte_iter = true;
+        let mut size = 0;
+        for &byte in self.as_bytes().iter() {
+            let b58_value = CHAR_ALPHABET_INDEX_MAP[byte as usize];
+            if b58_value == -1 {
+                return Err(FromBase58Error::InvalidCharacter);
+            }
+
+            if !encoding_flag {
+                if b58_value == 0 {
+                    final_bytes.push(0);
+                    continue;
+                }
+                encoding_flag = true;
+            }
+
+            let mut carry = b58_value as u32;
+            let mut i = 0;
+
+            while carry != 0 || i < size {
+                if first_byte_iter {
+                    raw_bytes.push(0);
+                    first_byte_iter = false;
+                } else {
+                    if let Some(raw_byte) = raw_bytes.get(i) {
+                        carry += *raw_byte as u32 * 58;
+                    } else {
+                        raw_bytes.push(0);
+                    }
+                }
+
+                raw_bytes[i] = (carry % 256) as u8;
+                carry /= 256;
+
+                i += 1
+            }
+
+            size = i;
+        }
+
+        raw_bytes.reverse();
```

Comment:
I would like to think later if there's another way to do the algorithm without this, but for now I think to change this is to optmize prematurely.

Reply:
It is possible. `bitcoin-core` and `trezor` implement it without reverse.
See https://github.com/bitcoin/bitcoin/blob/master/src/base58.cpp#L87 and https://github.com/trezor/trezor-crypto/blob/master/base58.c#L149, although it seems like the first implementation of Base58 encode in `bitcoin-core` actually used `reverse` and `Big Integer` (https://github.com/bitcoin/bitcoin/commit/b58be132c994b6f9b25cb4a702186ef96104953f).

I ended up not copying their code because I wanted to try implementing on my own, but we can copy their code later if performance starts to be an issue :)

We can also allocate memory upfront instead of calling `push` everytime, as they do.